frame/reward-distribution/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+        /// Distribute reward between beneficiaries.
+        ///
+        /// # Arguments
+        /// * `reward` - reward that will be split and distributed
+        ///
+        /// # Returns
+        /// * `weight` of the operation
+        ///
+        pub fn distribute_reward(reward: NegativeImbalanceOf<T>) -> Weight {
+            let distro_params = Self::reward_distribution_params();
+
+            // Pre-calculate balance which will be deposited for each beneficiary
+            let base_treasuy_balance = distro_params.base_treasury_percent * reward.peek();
+            let base_staker_balance = distro_params.base_staker_percent * reward.peek();
+            let dapps_balance = distro_params.dapps_percent * reward.peek();
+            let collator_balance = distro_params.collators_percent * reward.peek();
+
+            // This is part that's distributed between stakers and treasury
+            let adjustable_balance = distro_params.adjustable_percent * reward.peek();
+
+            // TVL percentage is fetched from outside this pallet
+            let tvl_percentage = T::DappsStakingTvlProvider::tvl_percentage();
+
+            // Calculate total staker and treasury reward balance
+            let adjustable_staker_part = if distro_params.ideal_dapps_staking_tvl.is_zero() {
+                adjustable_balance
+            } else {
+                tvl_percentage / distro_params.ideal_dapps_staking_tvl * adjustable_balance
+            };
+
+            let total_staker_balance = base_staker_balance + adjustable_staker_part;
+            let total_treasury_balance =
+                base_treasuy_balance + (adjustable_balance - adjustable_staker_part);
+
+            // Prepare imbalances
+            let (dapps_imbalance, remainder) = reward.split(dapps_balance);
+            let (stakers_imbalance, remainder) = remainder.split(total_staker_balance);
+            let (collator_imbalance, remainder) = remainder.split(collator_balance);
+            let (treasury_imbalance, remainder) = remainder.split(total_treasury_balance);
+
+            // TODO: remove this later, useful for testing atm
+            assert!(remainder.peek().is_zero());
+
+            // Payout beneficiaries
+            T::BeneficiaryPayout::treasury(treasury_imbalance);
+            T::BeneficiaryPayout::collators(collator_imbalance);
+            T::BeneficiaryPayout::dapps_staking(stakers_imbalance, dapps_imbalance);
+
+            // TODO: benchmark this code
+            100
+        }
+    }
+}
+
+// TODO (remove this)
+// Alternative to this approach would be to have some as configurable constants.
+// This approach, however, provides more flexibility since we can change them without
+// runtime upgrade.
+
+/// List of parameters used to calculate reward distribution portions for all the beneficiaries.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct RewardDistributionParams {
```

Comment:
Looks like RewardDistributionConfig is more suitable name for this struct

Reply:
Sure.